python-checks/src/main/java/org/sonar/python/checks/DictionaryDuplicateKeyCheck.java

```diff
@@ -0,0 +1,132 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.tree.DictionaryLiteral;
+import org.sonar.plugins.python.api.tree.DictionaryLiteralElement;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.KeyValuePair;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.NumericLiteral;
+import org.sonar.plugins.python.api.tree.StringElement;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+
+@Rule(key = "S5780")
+public class DictionaryDuplicateKeyCheck extends PythonSubscriptionCheck {
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.DICTIONARY_LITERAL, ctx -> {
+      DictionaryLiteral dictionaryLiteral = (DictionaryLiteral) ctx.syntaxNode();
+      Set<Integer> issueIndexes = new HashSet<>();
+      for (int i = 0; i < dictionaryLiteral.elements().size(); i++) {
+        if (!dictionaryLiteral.elements().get(i).is(Tree.Kind.KEY_VALUE_PAIR) || issueIndexes.contains(i)) {
+          continue;
+        }
+        Expression key = ((KeyValuePair) dictionaryLiteral.elements().get(i)).key();
+        List<Tree> duplicateKeys = findIdenticalKeys(i, dictionaryLiteral.elements(), issueIndexes);
+        if (!duplicateKeys.isEmpty()) {
+          PreciseIssue issue = ctx.addIssue(key, "Change or remove duplicates of this key.");
+          duplicateKeys.forEach(d -> issue.secondary(d, "Duplicate key"));
+        }
+      }
+    });
+  }
+
+  private List<Tree> findIdenticalKeys(int startIndex, List<DictionaryLiteralElement> elements, Set<Integer> issueIndexes) {
+    Expression key = ((KeyValuePair) elements.get(startIndex)).key();
+    List<Tree> duplicates = new ArrayList<>();
+    for (int i = startIndex + 1; i < elements.size() ; i++) {
+      if (!elements.get(i).is(Tree.Kind.KEY_VALUE_PAIR)) {
+        continue;
+      }
+      Expression comparedKey = ((KeyValuePair) elements.get(i)).key();
+      if (isSameKey(key, comparedKey)) {
+        issueIndexes.add(i);
+        duplicates.add(comparedKey);
+      }
+    }
+    return duplicates;
+  }
+
+  private boolean isSameKey(Expression key, Expression comparedKey) {
+    if (key.is(Tree.Kind.STRING_LITERAL) && comparedKey.is(Tree.Kind.STRING_LITERAL)) {
+      return areEquivalentStringLiterals((StringLiteral) key, (StringLiteral) comparedKey);
+    }
+    if (isANumber(key) && isANumber(comparedKey)) {
+      return areEquivalentNumbers(key, comparedKey);
+    }
+    return !key.is(Tree.Kind.CALL_EXPR) && CheckUtils.areEquivalent(key, comparedKey);
+  }
+
+  private boolean areEquivalentNumbers(Tree key, Tree comparedKey) {
+    return toBigDecimal(key).compareTo(toBigDecimal(comparedKey)) == 0;
```

Comment:
Good catch! A really brief comment about the surprising behavior of `.equals` might improve maintainability a tiny bit. Maybe something like "// note that it must be `.compareTo`, `.equals` is insufficient, because it takes scale into account"? Or maybe just link to some page that explains why it has to be `.compareTo`? (Optional)